Mathlib/Data/Real/IsNonarchimedean.lean

```diff
@@ -0,0 +1,138 @@
+  we have `f (n * a) ≤ (f a)`. -/
+theorem nmul_le {F α : Type*} [Ring α] [FunLike F α ℝ] [AddGroupSeminormClass F α ℝ]
+    {f : F} (hna : IsNonarchimedean f) {n : ℕ} {a : α} : f (n * a) ≤ f a := by
+  rw [← nsmul_eq_mul]
+  exact nsmul_le hna
+
+/-- If `f` is a nonarchimedean additive group seminorm on `α` and `x y : α` are such that
+  `f x ≠ f y`, then `f (x + y) = max (f x) (f y)`. -/
+theorem add_eq_max_of_ne {F α : Type*} [AddGroup α] [FunLike F α ℝ]
+    [AddGroupSeminormClass F α ℝ] {f : F} (hna : IsNonarchimedean f) {x y : α} (hne : f x ≠ f y) :
+    f (x + y) = max (f x) (f y) := by
+  let _ := AddGroupSeminormClass.toSeminormedAddGroup f
+  have := AddGroupSeminormClass.isUltrametricDist hna
+  simp_rw [← AddGroupSeminormClass.toSeminormedAddGroup_norm_eq] at hne ⊢
+  exact norm_add_eq_max_of_norm_ne_norm hne
+
+/-- Given a nonarchimedean additive group seminorm `f` on `α`, a function `g : β → α` and a finset
+  `t : Finset β`, we can always find `b : β`, belonging to `t` if `t` is nonempty, such that
+  `f (t.sum g) ≤ f (g b)` . -/
+theorem finset_image_add {F α β : Type*} [AddCommGroup α] [FunLike F α ℝ]
+    [AddGroupSeminormClass F α ℝ] [Nonempty β] {f : F} (hna : IsNonarchimedean f)
+    (g : β → α) (t : Finset β) :
+    ∃ b : β, (t.Nonempty → b ∈ t) ∧ f (t.sum g) ≤ f (g b) := by
+  let _ := AddGroupSeminormClass.toSeminormedAddCommGroup f
+  have := AddGroupSeminormClass.isUltrametricDist' hna
+  simp_rw [← AddGroupSeminormClass.toSeminormedAddCommGroup_norm_eq]
+  rcases t.eq_empty_or_nonempty with rfl|ht
+  · simp
+  have ht' : ∀ x ∈ t, ∀ y ∈ t, ∃ z ∈ t, ‖g z‖ = max ‖g x‖ ‖g y‖ := by
+    intros x hx y hy
+    rcases le_total ‖g x‖ ‖g y‖ with h|h
+    · exact ⟨y, hy, by simp [h]⟩
+    · exact ⟨x, hx, by simp [h]⟩
+  obtain ⟨x, hx, hx'⟩ := Finset.sup'_mem (α := ℝ) ((‖g ·‖) '' t) (by simpa [sup_eq_max] using ht') t
+    ht (‖g ·‖) (fun _ ↦ Set.mem_image_of_mem _)
+  refine ⟨x, fun _ ↦ by simpa using hx, ?_⟩
+  dsimp only at hx'
+  rw [hx']
+  exact ht.norm_sum_le_sup'_norm _
+
+/-- Given a nonarchimedean additive group seminorm `f` on `α`, a function `g : β → α` and a
+  nonempty finset `t : Finset β`, we can always find `b : β` belonging to `t` such that
+  `f (t.sum g) ≤ f (g b)` . -/
+theorem finset_image_add_of_nonempty {F α β : Type*} [AddCommGroup α] [FunLike F α ℝ]
+    [AddGroupSeminormClass F α ℝ] [Nonempty β] {f : F} (hna : IsNonarchimedean f)
+    (g : β → α) {t : Finset β} (ht : t.Nonempty) :
+    ∃ b : β, (b ∈ t) ∧ f (t.sum g) ≤ f (g b) := by
+  obtain ⟨b, hbt, hbf⟩ := finset_image_add hna g t
+  exact ⟨b, hbt ht, hbf⟩
+
+/-- Given a nonarchimedean additive group seminorm `f` on `α`, a function `g : β → α` and a
+  multiset `s : Multiset β`, we can always find `b : β`, belonging to `s` if `s` is nonempty,
+  such that `f (t.sum g) ≤ f (g b)` . -/
+theorem multiset_image_add {F α β : Type*} [AddCommGroup α] [FunLike F α ℝ]
```

Comment:
I need this as stated (I use it in a situation where there is not a bundled normed structure). In #17565 @pechersky mentioned that he was not able to generalize this proof.

Reply:
Why is any "generalisation" required? It'd be the exact same proof, just packaged slightly differently. Let me have a quick shot at it.

Reply:
Yes, I don't understand what the problem is here. Adding these lines to `Analysis/Normed/Group/Ultra.lean` works fine:
```
/--
Given a function `f : ι → M` and a finite set `t ⊆ ι`, we can always find `i : ι`, belonging to `t`
if `t` is nonempty, such that `‖∏ j in t, f j‖ ≤ ‖f i‖`.
-/
@[to_additive "Given a function `f : ι → M` and a finite set `t ⊆ ι`, we can always find `i : ι`,
belonging to `t` if `t` is nonempty, such that `‖∑ j in t, f j‖ ≤ ‖f i‖`."]
theorem exists_norm_finset_prod_le (t : Finset ι) [Nonempty ι] {f : ι → M} :
    ∃ i : ι, (t.Nonempty → i ∈ t) ∧ ‖∏ j in t, f j‖ ≤ ‖f i‖ := by
  rcases t.eq_empty_or_nonempty with rfl | ht
  · simp
  have ht' : ∀ x ∈ t, ∀ y ∈ t, ∃ z ∈ t, ‖f z‖ = max ‖f x‖ ‖f y‖ := by
    intros x hx y hy
    rcases le_total ‖f x‖ ‖f y‖ with h|h
    · exact ⟨y, hy, by simp [h]⟩
    · exact ⟨x, hx, by simp [h]⟩
  obtain ⟨x, hx, hx'⟩ := Finset.sup'_mem (α := ℝ) ((‖f ·‖) '' t) (by simpa [sup_eq_max] using ht') t
    ht (‖f ·‖) (fun _ ↦ Set.mem_image_of_mem _)
  refine ⟨x, fun _ ↦ by simpa using hx, ?_⟩
  dsimp only at hx'
  rw [hx']
  exact ht.norm_prod_le_sup'_norm _

/--
Given a function `f : ι → M` and a multiset `t : Multiset ι`, we can always find `i : ι`, belonging
to `t` if `t` is nonempty, such that `‖(s.map f).prod‖ ≤ ‖f i‖`.
-/
@[to_additive "Given a function `f : ι → M` and a multiset `t : Multiset ι`, we can always find
`i : ι`, belonging to `t` if `t` is nonempty, such that `‖(s.map f).sum‖ ≤ ‖f i‖`."]
theorem exists_norm_multiset_prod_le (s : Multiset ι) [Nonempty ι] {f : ι → M} :
    ∃ i : ι, (s ≠ 0 → i ∈ s) ∧ ‖(s.map f).prod‖ ≤ ‖f i‖ := by
  inhabit ι
  induction s using Multiset.induction_on with
  | empty => simp
  | @cons a t hM =>
      obtain ⟨M, hMs, hM⟩ := hM
      by_cases hMa : ‖f M‖ ≤ ‖f a‖
      · refine ⟨a, by simp, ?_⟩
        · rw [Multiset.map_cons, Multiset.prod_cons]
          exact le_trans (norm_mul_le_max _ _) (max_le (le_refl _) (le_trans hM hMa))
      · rw [not_le] at hMa
        rcases eq_or_ne t 0 with rfl|ht
        · exact ⟨a, by simp, by simp⟩
        · refine ⟨M, ?_, ?_⟩
          · simp [hMs ht]
          rw [Multiset.map_cons, Multiset.prod_cons]
          exact le_trans (norm_mul_le_max _ _) (max_le hMa.le hM)
```

Reply:
For `finset_image_add` there is also not an analogue in #17565 (instead, the proof inside the `IsNonarchimedean` file is modified). If you are looking into this, could you also have a look at this other lemma? Otherwise I do not see how to prove `add_pow_le` in the `[SeminormedRing R] [IsUltrametricDist R]` setting as you suggest below (and I do need that lemma for a proof in BGR).

Reply:
The two lemmas in my last message are reformulations of both `finset_image_add` and `multiset_image_add`. More precisely, I gave the proofs of the product versions, but the `@[to_additive]` decorator creates from `exists_norm_finset_prod_le` a lemma `exists_norm_finset_sum_le` with statement
```lean
IsUltrametricDist.exists_norm_finset_sum_le.{u_1, u_2} {M : Type u_1} {ι : Type u_2} [SeminormedAddCommGroup M]
  [IsUltrametricDist M] (t : Finset ι) [Nonempty ι] {f : ι → M} : ∃ i, (t.Nonempty → i ∈ t) ∧ ‖∑ j ∈ t, f j‖ ≤ ‖f i‖
```
which is exactly your `finset_image_add` written in terms of bundled norms.

Reply:
Sorry, my browser had not updated, so I had not seen your reply before I wrote my comment.

Reply:
As for the reason for the design:

In the construction of the spectral norm, I use the following lemma:
```lean
theorem isNonarchimedean_multiset_powerset_image_add {f : AlgebraNorm K L}
    (hf_na : IsNonarchimedean f) (s : Multiset L) (m : ℕ) :
    ∃ t : Multiset L, card t = card s - m ∧ (∀ x : L, x ∈ t → x ∈ s) ∧
      f (Multiset.map Multiset.prod (powersetCard (card s - m) s)).sum ≤ f t.prod := by
```
The proof of this lemma uses `multiset_image_add`; here `m` could be equal to `card s`, in which case `t` is empty and I would not be able to apply that lemma with a nonemptiness hypothesis. 

`isNonarchimedean_multiset_powerset_image_add` does get applied in the case of `m = card s` ( see the proof of  [max_root_norm_eq_spectral_value](https://github.com/mariainesdff/LocalClassFieldTheory/blob/4efe76367694bdaa74610e9603c37d80eb772d62/LocalClassFieldTheory/FromMathlib/SpectralNorm.lean#L725)), and it allows me to not have to split the proof into the cases `m < card s` and `m = card s`.

Reply:
Makes sense. Is it OK with you if I push to your PR branch to add the two lemmas above?

Reply:
Yes, thank you!